components/authz-service/storage/v2/postgres/postgres.go

```diff
@@ -950,6 +950,211 @@ func (p *pg) insertRoleWithQuerier(ctx context.Context, role *v2.Role, q Querier
 	return nil
 }
 
+/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+/* * * * * * * * * * * * * * * * * *    Rules    * * * * * * * * * * * * * * * * * * * */
+/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+
+func (p *pg) CreateRule(ctx context.Context, rule *v2.Rule) (*v2.Rule, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	tx, err := p.db.BeginTx(ctx, nil)
+	if err != nil {
+		return nil, p.processError(err)
+	}
+
+	row := tx.QueryRowContext(ctx,
+		`INSERT INTO iam_project_rules (id, project_id, name, type) VALUES ($1, $2, $3, $4) RETURNING db_id;`,
+		rule.ID, rule.ProjectID, rule.Name, rule.Type.String())
+	var ruleDbID string
+	if err := row.Scan(&ruleDbID); err != nil {
+		return nil, p.processError(err)
+	}
+
+	for _, condition := range rule.Conditions {
+		_, err := tx.ExecContext(ctx,
+			`INSERT INTO iam_rule_conditions (rule_id, value, attribute, operator) VALUES ($1, $2, $3, $4);`,
+			ruleDbID, condition.Value, condition.Attribute.String(), condition.Operator.String(),
+		)
+		if err != nil {
+			return nil, p.processError(err)
+		}
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return nil, storage_errors.NewErrTxCommit(err)
+	}
+
+	// Currently, we don't change anything from what is passed in.
+	return rule, nil
+}
+
+func (p *pg) UpdateRule(ctx context.Context, rule *v2.Rule) (*v2.Rule, error) {
+	projectsFilter, err := projectsListFromContext(ctx)
+	if err != nil {
+		return nil, p.processError(err)
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	tx, err := p.db.BeginTx(ctx, nil)
+	if err != nil {
+		return nil, p.processError(err)
+	}
+
+	row := tx.QueryRowContext(ctx,
+		`UPDATE iam_project_rules SET (name, type) = ($1, $2)
+			WHERE id = $3 AND projects_match_for_rule(project_id, $4) RETURNING db_id, project_id`,
+		rule.Name, rule.Type.String(), rule.ID, pq.Array(projectsFilter))
```

Comment:
Please see https://chefio.atlassian.net/browse/A2-789. I think you were around when we discussed that card. The piece missing here is to add a `RowsAffected` check after the operation to have an explicit check for returning a not found for project filtering.
(Same thing on get operation--you've already got it on delete.)

Reply:
It actually behaves correctly (process error returns ErrNotFound), but I can add the check to make it more explicit.

Reply:
We need it in delete because I'm doing an `ExecContext`. In get and update, we are using a `QueryRowContext` which will be converted to `storage_errors.ErrNotFound` correctly by `processError`. I have test coverage that indicates the correct thing is happening.

Reply:
Again, per A2-789, we agreed as a team that we want to be explicit on all-the-things (get, update, delete). If you want to re-open the discussion, that's cool, though.